### PR TITLE
Message when too many results

### PR DESCRIPTION
**Pull request by RenataMuellerC2C**

### CODE REVIEW

File: contribs/gmf/src/query/grid.scss [hunk 1]
```diff
@@ -100,3 +100,9 @@ $table-height: $grid-height - $app-margin - 2 * $map-tools-size;
 .gmf-query-grid-active main {
   height: calc(100% - #{$grid-height});
 }
+
+.gmf-query-grid-too-many {
+  color: #ff0000;
```
Comment (ger-benjamin) [inline contribs/gmf/src/query/grid.scss:105]: Please remove this `red` and use the bootstrap class `text-danger` to have the same red as other "error" messages.
Also, is red the right color ? perhaps a `text-warning` (yellow) is better ?

Suggestion: It looks better in italic to me.

File: contribs/gmf/src/query/gridComponent.js [hunk 2]
```diff
@@ -478,18 +478,26 @@ QueryGridController.prototype.getMergedSource_ = function (source, mergedSources
...
+      newRequest = !source.requestPartners.some((label) => mergeSource.mergeComposants.includes(label));
+    }
+    mergeSource.mergeComposants.push(source.label);
   } else {
     mergeSource = {
       features: [],
       id: mergeSourceId,
       label: mergeSourceId,
-      limit: this.maxResults,
+      limit: 0,
```
Comment (ger-benjamin) [inline contribs/gmf/src/query/gridComponent.js:497]: Are you sure of this change ?

File: contribs/gmf/src/query/gridComponent.js [hunk 3]
```diff
@@ -500,20 +508,24 @@ QueryGridController.prototype.getMergedSource_ = function (source, mergedSources
   });
 
   // if one of the source has too many results, the resulting merged source will
-  // also be marked with `tooManyResults` and will not contain any features.
+  // also be marked with `tooManyResults`.
   mergeSource.tooManyResults = mergeSource.tooManyResults || source.tooManyResults;
-  if (mergeSource.tooManyResults) {
+  /**if (mergeSource.tooManyResults) {
```
Comment (ger-benjamin) [inline contribs/gmf/src/query/gridComponent.js:513]: to remove ?

File: src/query/Querent.js [hunk 4]
```diff
@@ -398,6 +394,53 @@ export class Querent {
...
+   */
+  handleWFSQueryResult_(dataSources, maxFeatures, totalFeatureCount, wfs, response) {
+    /** @type {QuerentResult} */
+    const hash = {};
+    // For combination of layers the service gives some less than the maxFeatures allowed, so recompute the effective obtained number
+    const formatWFS = new olFormatWFS();
+    const limit = formatWFS.readFeatures(response.data).length;
+    const tooManyFeatures = totalFeatureCount > limit;
+    /** @type {string[]} */
+    const types = [];
```
Comment (ger-benjamin) [inline src/query/Querent.js:418]:
```suggestion
    const datasourceNames = [];
```
for clarity ?

File: src/query/Querent.js [hunk 5]
```diff
@@ -722,26 +759,28 @@ export class Querent {
       //
       //     If we do not need to count features first, then proceed with
       //     an normal WFS GetFeature request.
-      const getFeatureDefer = this.q_.defer();
+
+      /*const getFeatureDefer = this.q_.defer();
```
Comment (ger-benjamin) [inline src/query/Querent.js:763]: Remove me ?

File: src/query/Querent.js [hunk 6]
```diff
@@ -722,26 +759,28 @@ export class Querent {
...
-          {
-            resultType: 'hits',
-          },
-          getFeatureCommonOptions
-        );
-        const featureCountXml = wfsFormat.writeGetFeature(getCountOptions);
-        const featureCountRequest = xmlSerializer.serializeToString(featureCountXml);
-        const canceler = this.registerCanceler_();
-        countPromise = this.http_
+      //let countPromise;
```
Comment (ger-benjamin) [inline src/query/Querent.js:770]: remove me ?

File: src/query/Querent.js [hunk 7]
```diff
@@ -751,56 +790,42 @@ export class Querent {
             if (!dataSources[0].wfsFormat) {
               throw new Error('Missing wfsFormat');
             }
-            const meta = dataSources[0].wfsFormat.readFeatureCollectionMetadata(response.data);
-            if (!meta) {
-              throw new Error('Missing meta');
+
+            const metadata = dataSources[0].wfsFormat.readFeatureCollectionMetadata(response.data);
```
Comment (ger-benjamin) [inline src/query/Querent.js:794]: Do we always have metadata ? the previous `if` statement checked that

File: src/statemanager/WfsPermalink.js [hunk 8]
```diff
@@ -42,6 +42,8 @@ import olFormatWFS from 'ol/format/WFS.js';
  * @property {number} [totalFeatureCount] If `tooManyResults` is `true`, this contains the total number of
  *    features.
  * @property {string} [identifierAttributeField] The field to use to identify the feature.
+ * @property {string[]} [mergeComposants] The original types merged in merged result.
+ * @property {string[]} [requestPartners] The types of the request (which are summed in totalFeatureCount).
```
Comment (ger-benjamin) [inline src/statemanager/WfsPermalink.js:46]: Could you please search `types` in this PR and looks if you can use a better name. `Type` is not clear to me.
If this name is used somewhere else, please keep it for coherence.

**Review by ger-benjamin — COMMENTED** [review]
Works well. Some comments

File: contribs/gmf/src/query/gridComponent.js (hunk 2 above)
Comment (RenataMuellerC2C) [reply contribs/gmf/src/query/gridComponent.js]: The number of results is not always the given maxResults. If the merged sources come from different request each source has the maxResults which have to added. Also (see bug https://jira.camptocamp.com/browse/GSGMF-1457) the number of results obtained not always corresponds to maximum number of results allowed. Therefore the limit is updated later.

File: contribs/gmf/src/query/gridComponent.html [hunk 9]
```diff
@@ -43,83 +40,78 @@
...
-          <p><span translate>The results can not be displayed because the maximum number has been reached</span> ({{gridSource.source.limit}}).</p>
-          <p translate>Please refine your query.</p>
-        </div>
-      </div>
-    </div>
   </div>
+  <div class="container-fluid">
+  <div ng-show="!ctrl.pending && ctrl.getActiveGridSource() && ctrl.getActiveGridSource().configuration !== null" class="row">
+    <div class="col-md-5 my-auto" translate>
+      <span ng-if="ctrl.hasActiveSourceTooManyResults()" class="gmf-query-grid-too-many text-warning">Only {{ctrl.getActiveGridSourceLimit()}} of {{ctrl.getActiveGridSourceTotalFeatures()}} results displayed, as the maximum number is reached. Please refine your query. </span>
```
Comment (ger-benjamin) [inline contribs/gmf/src/query/gridComponent.html:49]: I think the translation will not work like that (with the translate in the parent markup).
You should have to do this in you case I think:
`{{'Only' | translate}} {{ctrl.getActiveGridSourceLimit()}} {{'of' | translate}} {{ctrl.getActiveGridSourceTotalFeatures()}} {{'results displayed, as the maximum number is reached. Please refine your query.' | translate}}`

File: contribs/gmf/src/query/gridComponent.js (hunk 2 above)
Comment (ger-benjamin) [reply contribs/gmf/src/query/gridComponent.js]: ok, thanks

**Comment by ger-benjamin** [comment]
One new comment, then it should be good

**Review by ger-benjamin — APPROVED** [review]

**Pull request merged**